<html><head>
<meta http-equiv="content-type" content="text/html; charset=UTF-8">
<link href="styles/globalstyle.css" type="text/css" rel="stylesheet"/>
<link title="medium" href="styles/two.css" type="text/css" rel="stylesheet"/>
<script src="scripts/main.js" type="text/javascript" language="javascript"></script>
<title>Section 12.1. BioPerl and the Bio::Graphics Module</title>
</head><body>
<DIV class=h3 style="MARGIN-TOP: 0px; PADDING-BOTTOM: 1em; WIDTH: 100%; PADDING-TOP: 0px"><SPAN>Growing Beautiful Code in BioPerl</SPAN><SPAN> &gt; BioPerl and the Bio::Graphics Module</SPAN></DIV>
<DIV></DIV>
<DIV>
<DIV><A name=growing_beautiful_code_in_bioperl></A>
<H2 class=docChapterTitle id=title-ID0EBGDK>12. Growing Beautiful Code in BioPerl</H2>
<P class=docText><SPAN class=docEmphasis><A name="Lincoln Stein"></A>Lincoln Stein</SPAN> <A name=idx-CHP-12-0855></A></P>
<P class=docText><SPAN class=docEmphSmaller><A name="biology has"></A>In the past decade, biology has blossomed into an information science</SPAN><A name="biologists with"></A>. New technologies provide biologists with unprecedented windows into the intricate processes going on inside the cells of animals and plants. DNA sequencing machines allow the rapid readout of complete genome sequences; microarray technologies give snapshots of the complex patterns of gene expression in developing organisms; confocal microscopes produce 3-D movies to track changes in cell architecture as precancerous tissues turn malignant.</P>
<P class=docText><A name="technologies routinely"></A>These new technologies routinely generate terabytes of data, all of which must be filtered, stored, manipulated and data-mined. The application of computer science and software engineering to the problems of biological data management is called <SPAN class=docEmphasis>bioinformatics</SPAN>.<A name=idx-CHP-12-0856></A></P>
<P class=docText><A name="in many"></A>Bioinformatics is similar in many ways to software engineering on Wall Street. Like software engineers in the financial sector, bioinformaticians need to be fleet of foot: they have to get applications up and running quickly with little time for requirement analysis and design. Data sets are large and mutable, with shelf lives measured in months, not years. For this reason, most bioinformatics developers favor agile development techniques, such as eXtreme Programming, and toolkits that allow for rapid prototyping and deployment. As in the financial sector, there is also a strong emphasis on data visualization and pattern recognition.</P><A name=bioperl_and_the_biographics_module></A>
<H3 class=docSection1Title id=-100000>12.1. BioPerl and the Bio::Graphics Module</H3>
<P class=docText><A name="bioinformaticians is"></A>One of the rapid development toolkits developed by and for bioinformaticians is BioPerl, an extensive open source library of reusable code for the Perl programming language. BioPerl provides modules to handle most common bioinformatics problems involving DNA and protein analysis, the construction and analysis of evolutionary trees, the interpretation of genetic data, and, of course, <A name=idx-CHP-12-0857></A>genome sequence analysis.<A name=idx-CHP-12-0858></A><A name=idx-CHP-12-0859></A><A name=idx-CHP-12-0860></A></P>
<P class=docText><A name="a software"></A>BioPerl allows a software engineer to rapidly create complex pipelines to process, filter, analyze, integrate, and visualize large biological datasets. Because of its extensive testing by the open source community, applications built on top of BioPerl are more likely to work right the first time, and because Perl interpreters are available for all major platforms, applications written with BioPerl will run on Microsoft Windows, Mac OS X, Linux, and Unix machines.</P>
<P class=docText><A name="map rendering"></A>This chapter discusses Bio::Graphics, BioPerl's genome map rendering module. The problem it addresses is how to visualize a genome and its <A name=idx-CHP-12-0861></A><A name="set of"></A>annotations. A genome consists of a set of <A name=idx-CHP-12-0862></A><A name="DNA sequences"></A>DNA sequences, each a string of the letters [A,G,C,T], which are <A name=idx-CHP-12-0863></A>nucleotides, also known as <A name=idx-CHP-12-0864></A>base pairs, or <SPAN class=docEmphasis>bp</SPAN><A name="DNA sequence"></A>. Some of the DNA sequence strings can be quite long: for example, the human genome consists of 24 DNA sequence strings, one each for chromosomes 1 through 22, plus the X and Y chromosomes. The longest of these, chromosome 1, is roughly 150,000,000 bp long (150 megabases).<A name=idx-CHP-12-0865></A></P>
<P class=docText><A name="DNA sequences"></A>Hidden inside these DNA sequences are multiple regions that play roles in cell metabolism, reproduction, defense, and signaling. For example, some sections of the chromosome 1 DNA sequence are <A name=idx-CHP-12-0866></A><A name="are "></A>protein-coding genes. These genes are "transcribed" by the cell into shorter <A name=idx-CHP-12-0867></A><A name="are transported"></A>RNA sequences that are transported from the cell nucleus into the cytoplasm; these RNA sequences are then translated into proteins responsible for generating energy, moving nutrients into and out of the cell, making the cell membrane, and so on. Other regions of the DNA sequence are <A name=idx-CHP-12-0868></A><A name="regulatory site"></A>regulatory in nature: when a regulatory protein binds to a specific regulatory site, a nearby protein-coding gene is "turned on" and starts to be transcribed. Some regions correspond to parasitic DNA: short regions of sequence that can replicate themselves semiautonomously and hitchhike around on the genome. Still other regions are of unknown significance; we can tell that they're important because they have been conserved among humans and other organisms across long evolutionary intervals, but we don't yet understand what they do.</P>
<P class=docText><A name="interpreting functionally"></A>Finding and interpreting functionally significant regions of the genome is called <SPAN class=docEmphasis>annotation</SPAN><A name="major focus"></A> and is now the major focus of the genome project. The annotation of a genome typically generates far more data than the raw DNA sequence itself. The whole <A name=idx-CHP-12-0869></A><A name="three gigabytes"></A>human genome sequence occupies just three gigabytes uncompressed, but its current annotation uses many terabytes (also see <A class=docLink href="095.the_design_of_the_gene_sorte.html#the_design_of_the_gene_sorte">Chapter 13</A>).</P><A name=example_of_biographics_output></A>
<H4 class=docSection2Title id=title-ID0E6KDK>12.1.1. Example of Bio::Graphics Output</H4>
<P class=docText><A name="To home"></A>To home in on "interesting" regions of the genome, biologists need to visualize how multiple annotations relate to each other. For example, a putative regulatory region is more likely to be functionally significant if it is spatially close to a protein-coding gene and overlaps with a region that is conserved between evolutionarily distant species.<A name=idx-CHP-12-0870></A><A name=idx-CHP-12-0871></A></P>
<P class=docText><A name="developers to"></A>Bio::Graphics allows bioinformatics software developers to rapidly visualize a genome and all its annotations. It can be used in a standalone fashion to generate a static image of a region in a variety of graphics formats (including PNG, JPEG, and SVG), or incorporated into a web or desktop application to provide interactive scrolling, zooming, and data exploration.</P>
<P class=docText><A class=docLink href="javascript:moveTo('a_sample_image_generated_by_biographics');">Figure 12-1</A><A name="of an"></A> gives an example of an image generated by Bio::Graphics. This image shows a region of the genome of <SPAN class=docEmphasis>C. elegans</SPAN><A name="illustrates several"></A> (a small soil-dwelling worm) that illustrates several aspects of a <A name=idx-CHP-12-0872></A><A name="by Bio"></A>typical image generated by Bio::Graphics. The image is divided vertically into a series of horizontal tracks. The top track consists of a scale that runs horizontally from left to right. The units are in kilobases ("k"), indicating thousands of DNA bases. The region shown begins at just before position 160,000 of the <SPAN class=docEmphasis>C. elegans</SPAN><A name="in toto"></A> chromosome I, and extends to just after position 179,000, covering 20,000 base pairs in toto. There are four annotation tracks, each of which illustrates increasingly complex visualizations.<A name=idx-CHP-12-0873></A></P><A name=a_sample_image_generated_by_biographics></A>
<P>
<CENTER>
<H5 class=docFigureTitle><A name="sample image"></A>Figure 12-1. A sample image generated by Bio::Graphics</H5><IMG id="" height=197 alt="" src="images/growing_beautiful_code_in_bioperl.0.png" width=500 border=0> </CENTER>
<P></P><BR>
<P class=docText><A name="brightly colored"></A>The original image is brightly colored, but has been reduced to grayscale here for printing. The simplest track is "cDNA for RNAi," which shows the positions of a type of experimental reagent that the research community has created for studying the regulation of <SPAN class=docEmphasis>C. elegans</SPAN><A name="annotation on"></A> genes. The image contains a single annotation on the right named <SPAN class=docEmphasis>yk247c7</SPAN><A name="It consists"></A>. It consists of a black rectangle that begins at roughly position 173,500 and extends to roughly 176,000. It corresponds to a physical piece of DNA covering this region, which a researcher can order from a biotech supply company and use experimentally to change the activity of the gene that overlaps it—in this case, F56C11.6.</P>
<P class=docText><A name="slightly more"></A>The "WABA alignments" track shows slightly more complex information. It visualizes quantitative data arising from comparing this part of the <SPAN class=docEmphasis>C. elegans</SPAN><A name="regions in"></A> genome to similar regions in a different worm. Regions that are highly similar are dark gray. Regions that are weakly similar are light gray. Regions of intermediate similarity are medium gray.</P>
<P class=docText><A name="shows continuously"></A>The "DNA/GC Content" track shows continuously variable quantitative information. This records the ratio of G and C nucleotides to A and T nucleotides across a sliding window of the nucleotide sequence. This ratio correlates roughly with the chances that the corresponding region of the genome contains a protein-coding gene.</P>
<P class=docText><A name="the most"></A>The "Genes" track contains the most complex data: the positions of two protein-coding genes. Each gene has an internal structure that indicates which parts encode protein (dark gray, but blue in the original) and which have other functions (lighter gray). Notice how the coding of the leftmost gene (F56C11.2) corresponds pretty well to the dark-gray, highly similar regions in the WABA alignment track; this is because the protein-coding regions of genes tend to be very strongly conserved across species.</P>
<P class=docText><A name="it is"></A>The gene named F56C11.6 is annotated with a function ("carboxylesterases"), indicating that it is related to a family of proteins responsible for a core part of carbon metabolism. In addition, it is shown with two alternative forms, indicating that it can encode more than one distinct protein. The two alternative forms are grouped together and given a distinct label. Notice that there are numerous alignments beneath this gene; this is a reflection that the gene belongs to a large family of related genes, and each related gene contributes to a different alignment.</P>
<P class=docText><A name="DNA nucleotide"></A>The actual DNA nucleotide sequence is not shown in this representation because it isn't physically possible to squeeze a line of 20,000 base pairs into 800 pixels. However, when viewing smaller segments of a genome, <A name=idx-CHP-12-0874></A><A name="actual letters"></A>Bio::Graphics can draw in the actual letters and ornament them (e.g., change their color or highlighting) to show the start and stop positions of interesting <A name=idx-CHP-12-0875></A>features.</P><A name=biographics_requirements></A>
<H4 class=docSection2Title id=title-ID0E1NDK>12.1.2. Bio::Graphics Requirements</H4>
<P class=docText><A name="take a"></A>The job of Bio::Graphics is to take a series of genome annotations (called <SPAN class=docEmphasis>features</SPAN><A name=in></A> in <A name=idx-CHP-12-0876></A><A name="a graphics"></A>BioPerl parlance) and output a graphics file formatted according to the programmer's specifications. Each feature has a <A name=idx-CHP-12-0877></A>start position, an <A name=idx-CHP-12-0878></A>end position, and a <A name=idx-CHP-12-0879></A><A name="not pointing"></A>direction (pointing left, pointing right, or not pointing in either direction). Features may be associated with other attributes such as a name, a description, and a numeric value. Features may also have an internal structure and contain <A name=idx-CHP-12-0880></A>subfeatures and sub-subfeatures.</P>
<P class=docText>When <A name=idx-CHP-12-0881></A><A name="I had"></A>designing this library, I had to address the following issues:</P><A name=idx-CHP-12-0882></A><A name=idx-CHP-12-0883></A><A name=idx-CHP-12-0884></A><A name=idx-CHP-12-0885></A><A name=idx-CHP-12-0886></A><A name=idx-CHP-12-0887></A><A name=idx-CHP-12-0888></A><A name=idx-CHP-12-0889></A><A name=idx-CHP-12-0890></A>
<DL class=docList>
<DT><BR>
<P><SPAN class=docPubcolor><SPAN class=docEmphasis>Open-ended nature of the problem</SPAN> </SPAN></P></DT>
<DD>
<P class=docList><A name=a></A>There are already a <A name=idx-CHP-12-0882></A><A name="While many"></A>large number of genome annotation types, and the number is growing daily. While many annotations can be drawn with simple rectangles of different colors, many of them—particularly the quantitative ones—can be quite complex to represent. Furthermore, different bioinformaticians may want to represent the same annotation type differently; <A name=idx-CHP-12-0883></A><A name="are many"></A>for example, there are many different ways of representing genes, each best suited for different circumstances.</P>
<P class=docList><A name="the open"></A>In order to accommodate the open-ended nature of this problem, I wanted to make it easy to add new visual representations to <A name=idx-CHP-12-0884></A><A name="representation should"></A>Bio::Graphics and to extend existing ones. Each representation should be highly configurable; for example, the programmer should be able to exercise exquisite control over the height, weight, boundary color, and fill color of even the simple rectangle. Furthermore, the programmer should be able to alter how each feature is rendered on a case-by-case basis.</P></DD>
<DT><BR>
<P><SPAN class=docPubcolor><SPAN class=docEmphasis><A name="Feature density"></A>Feature density</SPAN> </SPAN></P></DT>
<DD>
<P class=docList><A name="are very"></A>Some genomic features are very dense, whereas others are more sparse (compare the "Genes" and "WABA alignments" tracks in <A class=docLink href="javascript:moveTo('a_sample_image_generated_by_biographics');">Figure 12-1</A><A name="overlap spatially"></A>). Features can also overlap spatially. Sometimes you want overlapping features to partially obscure each other, and sometimes you'd like to practice collision control, shifting the overlapping features up or down vertically to make them distinct. To add to the problem, features can contain subfeatures that themselves overlap, so collision control and spatial layout need to work in a recursive fashion.<A name=idx-CHP-12-0885></A></P>
<P class=docList><A name="it would"></A>I thought it would be good if collision control could be activated and deactivated in a context-dependent manner. For example, if there are thousands of overlapping features in the region being displayed, collision control might cause a track to become unmanageably busy and should be turned off. The programmer should be able to control at what point this context-sensitive collision control kicks in, or override it entirely.</P></DD>
<DT><BR>
<P><SPAN class=docPubcolor><SPAN class=docEmphasis>Handling scale</SPAN> </SPAN></P></DT>
<DD>
<P class=docList><A name="be able"></A>I wanted Bio::Graphics to be able to draw pictures of a highly detailed 500 bp region of the genome, as well as a whole-genome view spanning 150 million nucleotides. To do this, the visual representation of features needs to change intelligently according to the current scale. At a scale of 500 bp, you can show the internal structure of a gene. At a scale of 1,000,000 bp, you can show only the start and end positions of the gene. At a scale of 100,000,000 bp, genes merge into a single black bar, and you should shift to a representation that shows the gene density as an intensity grayscale.<A name=idx-CHP-12-0886></A></P></DD>
<DT><BR>
<P><SPAN class=docPubcolor><SPAN class=docEmphasis>Useful for interactive web applications</SPAN> </SPAN></P></DT>
<DD>
<P class=docList><A name="interactive web"></A>I wanted Bio::Graphics to be suitable for the backend of an interactive web-based genome browser. In particular, I wanted end users to be able to click on the graphical renditions of features in order to bring up pull-down menus, link to other web pages, view tool tips, and so on. To accomplish this, Bio::Graphics had to be relatively fast so as to generate images in real time. It also had to keep track of the positions of each of its rendered features (and, if necessary, their subfeatures) in order to create an image map that could be passed back to the programmer.<A name=idx-CHP-12-0887></A></P></DD>
<DT><BR>
<P><SPAN class=docPubcolor><SPAN class=docEmphasis><A name="graphic formats"></A>Independence from graphic formats</SPAN> </SPAN></P></DT>
<DD>
<P class=docList><A name="useful for"></A>I wanted Bio::Graphics to be useful for generating screen-quality, low-resolution images suitable for embedding in a web page, as well as for generating high-resolution, publication-quality images. To do this, the part of Bio::Graphics that handled the logic of the layout had to be separate from the part that generated graphics. Ideally, it should be able to output both pixmap and vector graphics formats.<A name=idx-CHP-12-0888></A></P></DD>
<DT><BR>
<P><SPAN class=docPubcolor><SPAN class=docEmphasis>Independence from database schemes</SPAN> </SPAN></P></DT>
<DD>
<P class=docList><A name="many dozens"></A>The bioinformatics community has designed many dozens of database formats for managing genome annotation data, ranging from simple flat files to sophisticated relational databases. For maximum utility, I wanted to avoid tying <A name=idx-CHP-12-0889></A><A name="to any"></A>Bio::Graphics to any specific database scheme. It should be just as easy to invoke Bio::Graphics to render a genome region described by a flat file as to have it render a segment of a genome described in an Oracle database.<A name=idx-CHP-12-0890></A></P></DD></DL></DIV></DIV>
<p>&nbsp;</p><p>&nbsp;</p><!-- 仁·义 -->
</body></html>
