<html><head>
<meta http-equiv="content-type" content="text/html; charset=UTF-8">
<link href="styles/globalstyle.css" type="text/css" rel="stylesheet"/>
<link title="medium" href="styles/two.css" type="text/css" rel="stylesheet"/>
<script src="scripts/main.js" type="text/javascript" language="javascript"></script>
<title>Section 12.3. Extending Bio::Graphics</title>
</head><body>
<DIV class=h3 style="MARGIN-TOP: 0px; PADDING-BOTTOM: 1em; WIDTH: 100%; PADDING-TOP: 0px"><SPAN>Growing Beautiful Code in BioPerl</SPAN><SPAN> &gt; Extending Bio::Graphics</SPAN></DIV>
<DIV></DIV>
<DIV>
<DIV><A name=extending_biographics></A>
<H3 class=docSection1Title id=-100000>12.3. Extending Bio::Graphics</H3>
<P class=docText><A name="now look"></A>We'll now look at some of the Bio::Graphics extensions that were added after the initial release. This illustrates how code evolves in response to user input.<A name=idx-CHP-12-0987></A><A name=idx-CHP-12-0988></A></P><A name=supporting_web_developers></A>
<H4 class=docSection2Title id=title-ID0EVHDK>12.3.1. Supporting Web Developers</H4>
<P class=docText><A name="objectives of"></A>One of the objectives of Bio::Graphics was to support <A name=idx-CHP-12-0989></A><A name="of the"></A>interactive browsable views of the genome using web-based applications. My basic idea for this was that a CGI script would process a fill-out form indicating the genome to browse and a region to display. The script would make the database connection, process the user's request, find the region or regions of interest, pull out the features in the corresponding region, and pass them to Bio::Graphics. Bio::Graphics would render the image, and the CGI script would incorporate this data into an <TT>&lt;IMG&gt;</TT> tag for display.<A name=idx-CHP-12-0990></A></P>
<P class=docText><A name="thing missing"></A>The one thing missing from this picture was the ability to generate an image map for the generated image. An image map is necessary to support the user's ability to click on a glyph and get more information about it. <A name=idx-CHP-12-0991></A><A name="make it"></A>Image maps also make it possible to make tool tips appear when the user mouses over the glyph and to perform such dynamic HTML tasks as populating a pull-down menu when the user right-clicks on the glyph.</P>
<P class=docText><A name="single method"></A>To support image map generation, the original version of Bio::Graphics had a single method called <TT>boxes()</TT><A name="an array"></A>. This returned an array containing the glyph bounding rectangles, the features associated with each glyph, and the glyph objects themselves. To generate an image map, developers had to step through this array and generate the image map HTML manually.</P>
<P class=docText><A name="easy to"></A>Unfortunately, this was not as easy to do as I hoped it would be, judging from the number of user-support requests I received. So, after some experience writing my own Bio:: Graphics-based genome browser, I added an <TT>image_and_map()</TT><A name="code fragment"></A> method to Bio::Graphics:: Panel. Here is a code fragment that illustrates how to use this method:<A name=idx-CHP-12-0992></A></P>
<DIV class=codeSegmentsExpansionLinks>Code View: <SPAN>Scroll</SPAN> / <A href="javascript:expandCodeSegments()">Show All</A></DIV><PRE class=preFixedWidth>
	$panel = <A name=idx-CHP-12-0993></A>Bio::Graphics::Panel-&gt;new(@args);
	$panel-&gt;add_track(@args);
	$panel-&gt;add_track(@args);
	...

	($url,$map,$mapname) = $panel-&gt;image_and_map(
	                                -root =&gt; '/var/www/html',
	                                -url  =&gt; '/images',
	                                -link =&gt; sub {
	                                      my $feature = shift;
	                                      my $name = $feature-&gt;display_name;
	                                      return "http://www.google.com/search?q=$name";
	                                 }
	print "&lt;H2&gt;My Genome&lt;/H2&gt;";
	print "&lt;IMG SRC='$url' USEMAP='#$mapname' BORDER='0' /&gt;";
	print $map;


					    </PRE><BR>
<P class=docText><A name="We set"></A>We set up a Panel and add tracks to it as before. We then call <TT>image_and_map()</TT> with three argument/value pairs. The <TT>-root</TT><A name="physical location"></A> argument gives the physical location of the web server's document root—the place where the tree of HTML files starts. <TT>-url</TT><A name="the document"></A> indicates where, relative to the document root, Bio::Graphics-generated images should be stored. The <TT>-link</TT><A name="a callback"></A> argument is a callback that Bio::Graphics invokes to attach a clickable link to a glyph. In this case, we recover the feature object from the callback's argument list, get the feature's human-readable name by calling <TT>display_name()</TT><A name="Google search"></A>, and generate a Google search link. Several other <TT>image_and_map()</TT><A name="image map"></A> options can be used to customize and extend the resulting image map.</P>
<P class=docText><A name="generates the"></A>The method generates the image and stores it into the filesystem at the location indicated by <TT>-root</TT> and <TT>-url</TT>—in this case, <SPAN class=docEmphasis>/var/www/html/images</SPAN><A name="result list"></A>. It then returns a three-member result list consisting of a URL for the generated image, the HTML for the image map, and the name of the image map for use in the <TT>&lt;IMG&gt;</TT><A name="appropriate fragments"></A> tag. We then simply print the appropriate fragments of HTML to use the image and its map.</P>
<P class=docText><A name="To date"></A>To date there are two <A name=idx-CHP-12-0994></A>web-based <A name=idx-CHP-12-0995></A><A name="on Bio"></A>genome browsers based on Bio::Graphics. The one that I wrote, called <A name=idx-CHP-12-0996></A>GBrowse (<A class=docLink href="http://www.gmod.org/gbrowse" target=_blank>http://www.gmod.org/gbrowse</A><A name="used to"></A>), is now widely used to display a large number of genomes ranging from bacteria to man. However, it was written in 2002 before Ajax-based asynchronous page refreshes were invented; one moves along the genome by clicking arrow buttons and waiting for the screen to reload. A new browser that is currently in prototype stage, the <A name=idx-CHP-12-0997></A><A name="Genome Browser"></A>Ajax Generic Genome Browser (<A class=docLink href="http://genome.biowiki.org/gbrowse" target=_blank>http://genome.biowiki.org/gbrowse</A><A name="for the"></A>), provides Google Maps-style functionality for the genome. To navigate, one simply grabs the view and slides it.</P><A name=supporting_publication-quality_images></A>
<H4 class=docSection2Title id=title-ID0EMLDK>12.3.2. Supporting Publication-Quality Images</H4>
<P class=docText><A name="I designed"></A>Another original requirement was support for multiple graphics formats. To satisfy this, I designed Bio::Graphics to use the Perl <A name=idx-CHP-12-0998></A><A name="for its"></A>GD library for its low-level graphics calls. This library, which is based on Tom Boutell's <SPAN class=docEmphasis>libgd</SPAN> (<A class=docLink href="http://www.libgd.org/" target=_blank>http://www.libgd.org</A><A name="variety of"></A>), generates pixmap-based images in a variety of formats, including PNG, JPEG, and GIF.<A name=idx-CHP-12-0999></A><A name=idx-CHP-12-1000></A></P>
<P class=docText><A name="maintains a"></A>The Panel object creates and maintains a GD graphics object internally and passes this object to each of its tracks' <TT>draw()</TT><A name="The tracks"></A> routines. The tracks, in turn, pass the GD object to their glyphs, and the glyphs to their subglyphs.</P>
<P class=docText>The <TT>Bio::Graphics::Panel-&gt;png()</TT><A name="a pass"></A> method is simply a pass-through to GD's <TT>png()</TT> method:<A name=idx-CHP-12-1001></A><A name=idx-CHP-12-1002></A></P><PRE>	# in Bio::Graphics::Panel
	sub png {
	   my $self = shift;
	   my $gd = $self-&gt;gd;
	   return $gd-&gt;png;
	}
</PRE><BR>
<P class=docText>The <TT>jpeg()</TT> and <TT>gif()</TT><A name="developer also"></A> methods are similar. The developer also has the option of recovering the raw GD object and calling its <TT>png()</TT> method:</P><PRE>	$gd = $panel-&gt;gd;
	print $gd-&gt;png;
</PRE><BR>
<P class=docText><A name="of making"></A>The advantage of making the internal GD object available to the public interface is that the developer can do additional things with the GD object, such as embedding it in a larger picture or manipulating its colors.</P>
<P class=docText><A name="my choice"></A>One consequence of my choice to use GD was that Bio::Graphics was originally limited to the generation of pixmap images. This problem was solved by Todd Harris when he wrote the Perl GD:::<A name=idx-CHP-12-1003></A>SVG module (<A name=idx-CHP-12-1004></A><A class=docLink href="http://toddot.net/projects/GD-SVG" target=_blank>http://toddot.net/projects/GD-SVG</A>). <A name=idx-CHP-12-1005></A>GD::SVG has the same API as GD, but generates <A name=idx-CHP-12-1006></A><A name="without loss"></A>Scalable Vector Graphics (SVG) images, which can be printed at high resolution without loss of detail, and manipulated in various image-drawing applications such as Adobe Illustrator.</P>
<P class=docText><A name="added support"></A>After I added support for GD::SVG in Bio::Graphics, it became possible to produce SVGs simply by passing an <TT>-image_class</TT> argument to the Panel constructor:</P><PRE>	$panel = Bio::Graphics::Panel-&gt;new(-length=&gt;1000,
	                                      -width=&gt;600,
	                                      -image_class =&gt; 'GD::SVG'
	                                     );
	$panel-&gt;add_track.... etc...
	print $panel-&gt;gd-&gt;svg;
</PRE><BR>
<P class=docText><A name="had to"></A>Internally, the only change I had to make to Bio::Graphics was to process the <TT>-image_class</TT><A name="indicated image"></A> option and to load the indicated image library. This allows for forward compatibility with new GD-compatible libraries. For example, if someone writes a GD::PDF that generates PDF-format graphic files, Bio::Graphics will be able to accommodate it.</P><A name=adding_new_glyphs></A>
<H4 class=docSection2Title id=title-ID0EEPDK>12.3.3. Adding New Glyphs</H4>
<P class=docText><A name="At the"></A>At the time it was first published, Bio::Graphics supported about a dozen simple glyphs, including rectangles, ovals, arrows, the gene glyph, and a glyph that draws protein and DNA sequences. Each of these glyphs had multiple configuration options, leading to a very large number of possible displays. However, this number was still finite, whereas the number of feature types on the genome is potentially infinite. Fortunately, it is relatively easy to add new glyphs, and over time, I and other BioPerl developers have added many new glyphs to <A name=idx-CHP-12-1007></A><A name="nearly "></A>Bio::Graphics. Currently there are nearly 70 glyph types, ranging from whimsical (a Star of David) to sophisticated (a ternary plot for comparing frequencies of sequence variants in multiple populations).<A name=idx-CHP-12-1008></A></P>
<P class=docText><A name="easily extend"></A>The ability to easily extend existing glyphs to create new ones is a valuable feature. I will illustrate this in <A class=docLink href="javascript:moveTo('the_hourglass_glyph');">Example 12-5</A><A name="you how"></A> by showing you how to create a new Glyph called <TT>hourglass</TT>.</P><A name=the_hourglass_glyph></A>
<H5 class=docExampleTitle id=title-ID0ECQDK>Example 12-5. The hourglass glyph</H5>
<P>
<TABLE cellSpacing=0 cellPadding=5 width=* border=1>
<TBODY>
<TR>
<TD><PRE>1 package Bio::Graphics::Glyph::hourglass;

2 use strict;
3 use base 'Bio::Graphics::Glyph::box';

4 sub draw_component {
5   my $self = shift;
6   my ($gd,$dx,$dy) = @_;
7   my ($left,$top,$right,$bottom) = $self-&gt;bounds($dx,$dy);

8   # draw the hourglass as a polygon
9   my $poly = GD::Polygon-&gt;new;
10   $poly-&gt;addPt($left,$top);
11   $poly-&gt;addPt($right,$bottom);
12   $poly-&gt;addPt($right,$top);
13   $poly-&gt;addPt($left,$bottom);
14   $poly-&gt;addPt($left,$top);
15   $gd-&gt;filledPolygon($poly,$self-&gt;bgcolor);
16   $gd-&gt;polygon($poly,$self-&gt;fgcolor);
17 }

18 1;
</PRE><BR></TD></TR></TBODY></TABLE></P>
<P class=docText><A name="an hourglass"></A>This glyph generates an hourglass (<A class=docLink href="javascript:moveTo('the_hourglass_glyph_a_twisted_version_of_the_standard_box_gl');">Figure 12-5</A><A name="must begin"></A>). The glyph starts by defining its package name, which by convention must begin with <TT>Bio::Graphics::Glyph::</TT><A name="declares that"></A> (line 1). It then declares that it is inheriting from <TT>Bio::Graphics::Glyph::box</TT><A name="glyph that"></A>, which is a simple glyph that draws a rectangle (line 3).<A name=idx-CHP-12-1009></A></P><A name=the_hourglass_glyph_a_twisted_version_of_the_standard_box_gl></A>
<P>
<CENTER>
<H5 class=docFigureTitle><A name="version of"></A>Figure 12-5. The hourglass glyph, a twisted version of the standard box glyph</H5><IMG id="" height=65 alt="" src="images/extending_biographics.0.png" width=500 border=0> </CENTER>
<P></P><BR>
<P class=docText><A name="The glyph"></A>The glyph then overrides the inherited <TT>draw_component()</TT> method (lines 4–17). The <TT>draw_ component()</TT><A name="by a"></A> method is called by a <TT>draw()</TT><A name="setting up"></A> method of Bio::Graphics::Glyph after setting up the drawing environment. The method receives the GD object along with horizontal and vertical coordinates indicating the position of the glyph relative to its enclosing glyph. We pass the relative coordinates to the inherited <TT>bounds()</TT><A name="convert them"></A> method to convert them into the absolute coordinates of the rectangle enclosing the glyph (line 7).</P>
<P class=docText><A name="draw the"></A>Now we actually draw the glyph. We create a polygon using GD's polygon library and add vertices corresponding to the top-left, bottom-right, top-right, and bottom-left corners of the hourglass (lines 9–14). We then pass the polygon object first to the GD object's <TT>filledPolygon()</TT><A name="then to"></A> method to draw the solid contents of the polygon (line 15), and then to the GD object's <TT>polygon()</TT><A name="draw the"></A> method to draw the outline of the hourglass (line 16). Notice our use of the inherited <TT>bgcolor()</TT> and <TT>fgcolor()</TT><A name="appropriate colors"></A> methods to get the appropriate colors to use for the fill and the outline.</P>
<P class=docText><A name="adding new"></A>This demonstrates the simplicity of adding new glyphs to <A name=idx-CHP-12-1010></A><A name="In many"></A>Bio::Graphics. In many cases, one can create a new glyph by inheriting from an existing glyph that does almost what one wants, and then modifying one or two methods to customize it.</P></DIV></DIV>
<p>&nbsp;</p><p>&nbsp;</p><!-- 仁·义 -->
</body></html>
